plugin/completion.py

```diff
@@ -1,36 +1,83 @@
 import sublime
 import sublime_plugin
-from .core.completion import parse_completion_response, format_completion
-from .core.configurations import is_supported_syntax
-from .core.documents import position_is_word
-from .core.edit import parse_text_edit
+
+from .core.protocol import Request, Range, InsertTextFormat
+from .core.settings import settings, client_configs
 from .core.logging import debug
-from .core.protocol import Request
+from .core.completion import parse_completion_response, format_completion
 from .core.registry import session_for_view, client_from_session, LSPViewEventListener
+from .core.configurations import is_supported_syntax
 from .core.sessions import Session
-from .core.settings import settings, client_configs
+from .core.edit import parse_text_edit
+from .core.views import range_to_region
 from .core.typing import Any, List, Dict, Tuple, Optional, Union
 from .core.views import text_document_position_params
 
 
-class CompletionState(object):
-    IDLE = 0
-    REQUESTING = 1
-    APPLYING = 2
-    CANCELLING = 3
+class LspSelectCompletionItemCommand(sublime_plugin.TextCommand):
+    def run(self, edit: Any, item: dict) -> None:
+        insert_text_format = item.get("insertTextFormat")
 
+        text_edit = item.get('textEdit')
+        if text_edit:
+            # insert the removed command completion item prefix
+            # so we don't have to calculate the offset for the textEdit range
+            for sel in self.view.sel():
```

Comment:
Regarding handling multiple cursors: it is much more complex. If we want to do multiple cursors, we would have to request for each cursor location completions, and then take the *intersection* of all results. The remaining completion items are valid for all cursor locations. What we do here is insufficient. So let's not pretend we support this feature.

Reply:
Support for multicursor `TextEdit`-s is more complex than I thought :)

Reply:
Sorry, this code here `for sel in self.view.sel():` 
is used to insert back the removed prefix, 
so the text edit range for the completion item is valid again.
it is not used to to support multicursor text edits.

Reply:
Thanks for the explanation :)